working/sealed-types/feature-specification.md

```diff
@@ -0,0 +1,196 @@
+# Sealed types
+
+Author: Bob Nystrom
+
+Status: In-progress
+
+Version 1.0
+
+This proposal specifies *sealed types*, which is core capability needed for
+[exhaustiveness checking][] of subtypes in [pattern matching][]. This proposal
+is a subset of the [type modifiers][] proposal. (We may wish to do all or parts
+of the rest of that proposal, but they aren't needed for pattern matching, so
+this proposal separates them out.) For motivation, see the previously linked
+documents.
+
+[exhaustiveness checking]: https://github.com/dart-lang/language/blob/master/working/0546-patterns/exhaustiveness.md
+
+[pattern matching]: https://github.com/dart-lang/language/blob/master/working/0546-patterns/patterns-feature-specification.md
+
+[type modifiers]: https://github.com/dart-lang/language/blob/master/working/type-modifiers/feature-specification.md
+
+## Introduction
+
+Marking a type `sealed` applies two restrictions:
+
+*   If it's a class, the type itself can't be directly constructed. The class is
+    implicitly `abstract`.
+
+*   All direct subtypes of the type must be defined in the same library. Any
+    types that directly implement, extend, or mixin the sealed type must be
+    defined in the library where the sealed type is defined.
+
+In return for those restrictions, sealed provides two useful properties for
+exhaustiveness checking:
+
+*   All of the direct subtypes of the sealed type can be easily found and
+    enumerated.
+
+*   Any concrete instance of the sealed type must also be an instance of at
+    least one of the known direct subtypes. In other words, if you match on a
+    value of the sealed type and you have cases for all of the direct subtypes,
+    the compiler knows those cases are exhaustive.
+
+### Open subtypes
+
+Note that it is *not* necessary for the subtypes of a sealed type to themselves
+be sealed or closed to subclassing or implementing. Given:
+
+```dart
+sealed class Either {}
+
+class Left extends Either {}
+class Right extends Either {}
+```
+
+Then this switch is exhaustive:
+
+```dart
+test(Either either) {
+  switch (either) {
+    case Left(): print('Left');
+    case Right(): print('Right');
+  }
+}
+```
+
+And this is still true even if some unrelated or unknown library contains:
+
+```dart
+class LeftOut extends Left {}
+```
+
+Or even:
+
+```dart
+class Ambidextrous implements Left, Right {}
+```
+
+The only property we need for exhaustiveness is that *all instances of the
+sealed type must have one of the direct subtypes in their inheritance chain.*
```

Comment:
This isn't right, I don't think.  It's neither enforced by this, nor needed.  I think you want to say something like "all instances of whose type are subtypes of the sealed type must have at least one of the direct subtypes in their super-interface graph."?

Reply:
Oops, yes. Reworded.